reps/2024-10-18-train-tune-api-revamp/2024-10-18-train-tune-api-revamp.md

```diff
@@ -0,0 +1,1194 @@
+
+Here’s an example of what job-level restoration looks like before and after, where the job is submitted with this `python main.py <UNIQUE_JOB_ID>` as the entrypoint:
+
+<table>
+<!-- Table Headers -->
+  <tr>
+    <th>Before</th>
+    <th>After</th>
+  </tr>
+  <tr>
+<td>
+
+```python
+# main.py
+
+def train_fn_per_worker(config):
+    checkpoint = ray.train.get_checkpoint()
+
+    ...
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("run_id", type=str)
+    args = parser.parse_args()
+
+    unique_run_name = args.run_id  # some UNIQUE job id
+    storage_path = "s3://my/bucket"
+
+    run_path = f"{storage_path}/{unique_run_name}"
+
+    if TorchTrainer.can_restore(run_path):
+        # Certain configs get loaded from a pkl file,
+        # and other parameters need to be re-specified.
+        trainer = TorchTrainer.restore(
+            run_path,
+            train_loop_per_worker=train_fn_per_worker,
+            datasets={...},
+        )
+    else:
+
+        trainer = TorchTrainer(
+            train_fn_per_worker,
+            ...,
+            run_config=ray.train.RunConfig(
+                storage_path=storage_path,
+                name=unique_run_name,
+            ),
+        )
+
+    trainer.fit()
+```
+
+</td>
+<td>
+
+```python
+# main.py
+
+def train_fn_per_worker(config):
+    checkpoint = ray.train.get_checkpoint()
+
+    ...
+
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("run_id", type=str)
+    args = parser.parse_args()
+
+    unique_run_name = args.run_id  # some UNIQUE job id
+    storage_path = "s3://my/bucket"
+
+    trainer = TorchTrainer(
+        train_fn_per_worker,
+        ...,
+        run_config=ray.train.RunConfig(
+            storage_path=storage_path,
+            name=unique_run_name,
+        ),
+    )
+    # If `{storage_path}/{unique_run_name}` already contains a previous run's output,
+    # populate the latest checkpoint, which can be retrieved via `ray.train.get_checkpoint`.
+    trainer.fit()
+
+```
+
+</td>
+  </tr>
+
+</table>
+
+⚠️ New training runs should be assigned a unique `name`. We recommend generating this unique name as a uuid before submitting the job. Then, upon any driver failures, a job retry would re-initialize the run with the same `name`, which will load the latest checkpoint information from persistent storage.
+
+#### `resume_from_checkpoint`
+
+❌ `Trainer(resume_from_checkpoint)` was a common source of confusion and will be deprecated. `resume_from_checkpoint` is meant to provide a starting checkpoint for a new run, rather than restore all state from the previous run. See the following example for a more intuitive way to accomplish the same thing:
+
+```py
+def train_fn_per_worker(config: dict):
+    # Equivalent behavior that is explicit and more flexible.
+    checkpoint = (
+        ray.train.get_checkpoint()
```

Comment:
Minor, but isn't `config.get("resume_from_checkpoint")` enough? Why do you also need `ray.train.get_checkpoint()`?

Reply:
`ray.train.get_checkpoint` is still needed if you want to support fault tolerance, in addition to a checkpoint to start fine-tuning from initially. Otherwise, if a node fails, the training progress always gets reset to the initial checkpoint, throwing away all finetuning progress.